providers/openfeature-provider-flagd/src/openfeature/contrib/provider/flagd/resolvers/grpc.py

```diff
@@ -26,19 +32,123 @@
+        channel_factory = grpc.secure_channel if config.tls else grpc.insecure_channel
+        channel = channel_factory(
+            f"{config.host}:{config.port}",
+            options=(("grpc.keepalive_time_ms", config.keep_alive_time),),
         )
-        self.channel = channel_factory(f"{self.config.host}:{self.config.port}")
-        self.stub = evaluation_pb2_grpc.ServiceStub(self.channel)
+        stub = evaluation_pb2_grpc.ServiceStub(channel)
+        return stub, channel
+
+    def initialize(self, evaluation_context: EvaluationContext) -> None:
+        self.connect()
 
     def shutdown(self) -> None:
+        self.active = False
         self.channel.close()
 
+    def connect(self) -> None:
+        self.active = True
+        self.thread = threading.Thread(
+            target=self.listen, daemon=True, name="FlagdGrpcServiceWorkerThread"
+        )
+        self.thread.start()
+
+        ## block until ready or deadline reached
+        timeout = self.deadline + time.time()
+        while not self.connected and time.time() < timeout:
+            time.sleep(0.05)
+        logger.debug("Finished blocking gRPC state initialization")
+
+        if not self.connected:
+            raise ProviderNotReadyError(
+                "Blocking init finished before data synced. Consider increasing startup deadline to avoid inconsistent evaluations."
+            )
+
+    def listen(self) -> None:
+        retry_delay = self.retry_backoff_seconds
+
+        call_args = (
+            {"timeout": self.streamline_deadline_seconds}
+            if self.streamline_deadline_seconds > 0
+            else {}
+        )
+        while self.active:
+            request = evaluation_pb2.EventStreamRequest()
+            try:
+                logger.debug("Setting up gRPC sync flags connection")
+                for message in self.stub.EventStream(request, **call_args):
+                    if message.type == "provider_ready":
+                        if not self.connected:
+                            self.emit_provider_ready(
+                                ProviderEventDetails(
+                                    message="gRPC sync connection established"
+                                )
+                            )
+                            self.connected = True
+                            # reset retry delay after successsful read
+                            retry_delay = self.retry_backoff_seconds
+
+                    elif message.type == "configuration_change":
+                        data = MessageToDict(message)["data"]
+                        self.handle_changed_flags(data)
+
+                    if not self.active:
+                        logger.info("Terminating gRPC sync thread")
+                        return
+            except grpc.RpcError as e:
+                logger.error(f"SyncFlags stream error, {e.code()=} {e.details()=}")
+                self.stub, self.channel = self._create_stub()
```

Comment:
```suggestion
                # re-create the stub if there's a connection issue - otherwise reconnect does not work as expected
                self.stub, self.channel = self._create_stub()
```